tests/all/component_model/bindgen/results.rs

```diff
@@ -0,0 +1,100 @@
+use super::engine;
+use anyhow::{anyhow, Error};
+use wasmtime::{
+    component::{Component, Linker},
+    Store,
+};
+
+mod empty_error {
+    use super::*;
+    wasmtime::component::bindgen!({
+        inline: "
+        world result-playground {
+            import imports: interface {
+                empty-error: func(a: float64) -> result<float64>
+            }
+
+            default export interface {
+                empty-error: func(a: float64) -> result<float64>
+            }
+        }",
+        // This doesn't actually do anything, just showing the macro parses it correctly:
+        trappable_error_type: { imports::e: TrappableE }
+    });
+
+    #[test]
+    fn run() -> Result<(), Error> {
+        let engine = engine();
+        let component = Component::new(
+            &engine,
+            r#"
+            (component
+                (import "imports" (instance $i
+                    (export "empty-error" (func (param "a" float64) (result (result float64))))
+                ))
+                (core module $libc
+                    (memory (export "memory") 1)
+                    (func (export "realloc") (param i32 i32 i32 i32) (result i32)
+                        unreachable)
+                )
+                (core instance $libc (instantiate $libc))
+                (core module $m
+                    (import "" "core_empty_error" (func (param f64) (result i32)))
+                    (export "core_empty_error" (func 0))
+                )
+                (core func $core_empty_error
+                    (canon lower (func $i "empty-error") (memory $libc "memory"))
+                )
+                (core instance $i (instantiate $m
+                    (with "" (instance (export "core_empty_error" (func $core_empty_error))))
+                ))
+                (func $f_empty_error
+                    (export "empty-error")
+                    (param "a" float64)
+                    (result (result float64))
+                    (canon lift (core func $i "empty_error") (memory $libc "memory"))
+                )
+            )
+        "#,
+        )?;
+
+        #[derive(Default)]
+        struct MyImports {}
+
+        impl imports::Imports for MyImports {
+            fn empty_error(&mut self, a: f64) -> Result<Result<f64, ()>, Error> {
```

Comment:
Oh I  thought the `Error<T>` was going to work which would avoid this result-of-result, but could you explain a bit more why it didn't work out?

Reply:
I just don't have any tests that actually use the functionality in yet.

We can't use `Result<f64, Error<()>>` in this case because Error<T> requires T impl std::error::Error. Additionally, all of this PR only supports substituting named types, and doesn't support substituting primitive types.

Reply:
Ah ok makes sense. I'm worried about the ergonomics of this, though. Are you thinking we'd special-case `-> result<T, rich-error>` in WIT to have a `Result<T, E>` return value here? For degenerate cases like `-> result<T>` with no error or primitive errors I'm fine with doing whatever.

Reply:
I put the tests in below that show we can special case a `interface i { ... result<t, e> }` to `Result<T, TrappableE>` when the user invokes the macro with `trappable_error_types: { i::e: TrappableE, ... }`